### PR TITLE
Update tensor.view_names / tensor.names_ API

### PR DESCRIPTION
**Pull request by zou3519**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #24306 Implement name inference for mm, addmm
* #24203 Implement name inference for t(), transpose(...)
* #24087 Name inference for softmax, log_softmax and Dimname overloads.
* #24028 Fix out= function semantics for named tensors.
* **#23973 Update tensor.view_names / tensor.names_ API**
* #23962 Rename set_names -> view_names, set_names_ -> names_
* #23804 Implement tensor.align_to(names), torch.align_tensors(t1, t2, ...)
* #24109 Named inference for contiguous(), bernoulli variants, and dropout.
* #24108 Implement name inference rule for empty_like, clone
* #24107 Add `names` argument to ones, rand, randn, zeros, full; fix empty

Without loss of generality, I describe the API for `tensor.view_names`.
`tensor.names_` has an analogous API.

`tensor.view_names(*names)` returns a view on tensor with named dims `names`.
`names` must be of length `tensor.dim()`; otherwise, if '*' is in `names`,
then it (known as the "glob") is expanded greedily to be equal to the
corresponding names from `tensor.names`.

For example,
```
>>> x = torch.empty(2, 3, 5, 7, names=('N', 'C', 'H', 'W'))
>>> x.view_names('*', 'height', 'width').names
('N', 'C', 'height', 'width')

>>> x.view_names('batch', '*', 'width').names
('batch', 'C', 'H', 'width')
```

tensor.view_names(**rename_map) returns a view on tensor that has
renamed dims as specified in the mapping `rename_map`.

For example,
```
>>> x = torch.empty(2, 3, 5, 7, names=('N', 'C', 'H', 'W'))
>>> x.view_names(W='width', H='height').names
('N', 'C', 'height', 'width')
```

These are different(!!!) from the C++ API, which only allows the
following:
- tensor.view_names(optional<DimnameList>)

C++ API parity for named tensors is not important right now; I am
punting that to the future.

Test Plan:
- [namedtensor ci]

Differential Revision: [D16710916](https://our.internmc.facebook.com/intern/diff/D16710916)

### CODE REVIEW

**Comment by nairbv** [comment]
Is this a view in the same sense that `view` creates a view of the original tensor?
e.g.:

```
>>> a=torch.ones(4)
>>> b=a.view(2,2)
>>> b[0,0]=2
>>> a
tensor([2., 1., 1., 1.])
```

If so is there a test that demonstrates this behavior? I don't see one but may have missed it. If this isn't the expected behavior then the name might still be confusing.

**Comment by zou3519** [comment]
> Is this a view in the same sense that `view` creates a view of the original tensor?
> e.g.:
> 
> ```
> >>> a=torch.ones(4)
> >>> b=a.view(2,2)
> >>> b[0,0]=2
> >>> a
> tensor([2., 1., 1., 1.])
> ```

Yes.

> If so is there a test that demonstrates this behavior? I don't see one but may have missed it. If this isn't the expected behavior then the name might still be confusing.

There is a test hidden inside all of the tests, but I'll make an explicit test case for this.

> If this is correct, I think I might prefer something like `named_view` that passively describes what's being returned rather than the verb form.

It sounds better but I'm worried that it diverges too much from its in-place version `.names_(...)`.

**Comment by nairbv** [comment]
I'm checking for any other `view` methods ... just a thought, would an overload of `view_as` make sense? i.e. `view_as(Tensor self, DimnameList dimNames)`

I also see `reshape` and `reshape_as` which seem to create views. Not sure if `rename` makes sense. Personally I'd prefer past-tense `renamed` to indicate that it's the returned value not the in-place value that will have the new names, though `renamed` could be inconsistent with `reshape`. Looks like the `reshape` name comes from numpy.

**Comment by zou3519** [comment]
> I'm checking for any other `view` methods ... just a thought, would an overload of `view_as` make sense? i.e. `view_as(Tensor self, DimnameList dimNames)`

This could work, let me go think about that some more.

> I also see `reshape` and `reshape_as` which seem to create views. Not sure if `rename` makes sense. Personally I'd prefer past-tense `renamed` to indicate that it's the returned value not the in-place value that will have the new names, though `renamed` could be inconsistent with `reshape`. Looks like the `reshape` name comes from numpy.

`tensor.renamed(...)` ? That is also reasonable. I'm liking this better than a `view_as` overload right now.

Aside from the naming issue, how does the rest of this PR look? We can always punt a full resolution to what the naming of the the functions should be down the road since we're still in an experimental phase.

**Review by nairbv — APPROVED** [review]
Approving, though we may want to re-name the re-name function again before release
